### PR TITLE
Add waiting for completion overload for Indexer.run() API

### PR DESCRIPTION
**Pull request by chaosrealm**
Add waiting for completion overload for `Indexer.run()` API

### CODE REVIEW

**Review by EvanBoyle — COMMENTED** [review]
Do we really need a second API for this? 

Should this just be a parameter instead? 

We could add a param ` waitForCompletion: boolean` that returns a promise of a different shape rather than adding another method. 

I guess my other question is what is the use case?

**Comment by chaosrealm** [comment]
> What is the use case?
When running an indexer, it's very common for people to want to know the result of the run - at least that's been my experience in Azure Search. That why I had the polling loop in the docs - but that it occurred to me that if it's going to be a common scenario, it's better for us to provide a convenience API for it instead of providing an ample opportunity for users to have bugs in the polling loop. 

> Do we really need a second API for this? Should this just be a parameter instead?
I considered that - it's possible in TypeScript to overload based on the type of return value, but I thought it would be easier to use and document if it's a separate API. Certainly, open to just making this an optional argument to run if you think that will be cleaner.

**Comment by jmoseley** [comment]
How long does it take for an average indexer to run to completion?

**Comment by EvanBoyle** [comment]
> it's very common for people to want to know the result of the run

We don't have any users of the SDK yet (or really indexers in general) so I'm skeptical of this. I don't think it is a bad feature to add, but indexers take ~60s for the smallest data sets, and could be 10+ minutes for large ones, so I'm not sure this will actually see a ton of usage among our current users. 

If we are going to add it, I'd prefer to just have one method for running an indexer, similar to how we have one method for chat/content gen that has a `stream: boolean` parameter.

**Comment by chaosrealm** [comment]
> How long does it take for an average indexer to run to completion?

I think there's going to be huge variance. So far most of our examples complete within a few seconds and "medium" examples complete within ~20 seconds, but this obviously will vary based on the data source and load on the indexer queue.

**Comment by chaosrealm** [comment]
I agree that we don't have SDK users yet - that equally applies to most other SDK APIs we've added. I really think this will be useful - right away, it would immediately let us simplify sample code in the docs. 

> If we are going to add it, I'd prefer to just have one method for running an indexer, similar to how we have one method for chat/content gen that has a stream: boolean parameter.

Sure I'll go that route. Will help me internalize how to do method overloads in TypeScript :) 

I'll come back to this PR in a bit, after dealing with the Delete Member API

File: src/indexers/hosted/indexer.ts
Comment (EvanBoyle) [inline src/indexers/hosted/indexer.ts:69]: What is the rationale for a timeout vs just a simple boolean?

Comment (chaosrealm) [reply src/indexers/hosted/indexer.ts:69]: I started with a Boolean AND a timeout - we definitely need a timeout because deciding how long to wait is up to the user, even if we provide a default. Then I realized that we can use the presence of a timeout to obviate the need for a Boolean, we just need one knob instead of two

Comment (chaosrealm) [reply src/indexers/hosted/indexer.ts:69]: I guess we'll also need to document that an Error will be thrown if timeout has been exceeded without indexer completing - but that's even more true if we provide a default timeout implicitly

Comment (EvanBoyle) [reply src/indexers/hosted/indexer.ts:69]: Why can't the default timeout be no timeout?

Comment (chaosrealm) [reply src/indexers/hosted/indexer.ts:69]: Meaning we wait indefinitely long?

Comment (chaosrealm) [reply src/indexers/hosted/indexer.ts:69]: I instinctively didn't even consider this option because I've been conditioned on providing a timeout for any waiting API, but given huge variation in indexer execution time, this may be a good option!

Is -1 also considered a standard for "infinite" in TS world?

Comment (EvanBoyle) [reply src/indexers/hosted/indexer.ts:69]: For not let's not even bother worrying about timeouts until someone complains about it. Let's assume we don't timeout. We can add a timeout parameter in the future if needed, but let's avoid the additional complexity for now.

Comment (jmoseley) [reply src/indexers/hosted/indexer.ts:69]: The user can definitely make a timeout themselves a number of ways (eg Promise.race).

Comment (chaosrealm) [reply src/indexers/hosted/indexer.ts]: This is now exactly in the shape we discussed, afaict. Going to merge this tomorrow morning unless I get more feedback.

**Review by EvanBoyle — APPROVED** [review]

File: src/indexers/hosted/indexer.ts
Comment (EvanBoyle) [reply src/indexers/hosted/indexer.ts]: (reminder we need to turn on approval requirements for this repo and all others)

**Review by jmoseley — APPROVED** [review]

File: src/indexers/hosted/indexer.ts
Comment (chaosrealm) [reply src/indexers/hosted/indexer.ts]: Will turn them on for this and website repo. We've had it on for the cortex repo for a while.

**Pull request merged**
